Adding performance and data science concepts (#8893)
conda/conda: docs/source/user-guide/concepts/environments.rst

|

Other conda environments usually contain the same subdirectories
as the default environment.

Virtual environments
====================

A virtual environment is a tool that helps to
keep dependencies required by different projects
separate by creating isolated Python virtual environments
for them.

> Users can select a Python virtual environment, such as

Reviewer: Users can create virtual environments using one of several tools...

Pipenv or Poetry, or a conda virtual

> environment. Some characteristics of the two:

Reviewer: ```suggestion
environment. Pipenv and Poetry are based around Python's built-in venv library, whereas conda has its own notion of virtual environments that is lower-level (python itself is a dependency provided in conda environments).  Some other traits are:
```

.. list-table::
   :widths: 20 40 40
   :header-rows: 1

   * - 
     - Python virtual environment
     - Conda virtual environment
   * - **Libraries**
     - Statically link, vendor libraries in wheels,
       or use apt/yum/brew/etc.
     - Install system-level libraries as conda dependencies
   * - **System**
     - Depend on base system install of Python
     - Python is independent from system
   * - **Extending environment**
     - Extend environment with pip
     - Extended environment with conda or pip
   * - **Non-Python dependencies**
     -
     - Manages non-Python dependencies (R, Perl,
       arbitrary executables)
   * - **Tracking dependencies**
     -
     - Tracks binary dependencies explicitly

|

> Why use Python virtual environments

Reviewer: ```suggestion
Why use venv-based virtual environments
```

===================================
- You prefer their workflow or spec formats
- You prefer to use the system Python and libraries

> - You prefer packages that come more directly from the project maintainers

Reviewer: ```suggestion
- Ryour project maintainers only publish to PyPI, and you prefer packages that come more directly from the project maintainers, rather than someone else providing builds based on the same code
```

Why use conda virtual environments?
===================================

- You want control over binary compatibility choices

> - You want to utilize newer language standards

Reviewer: ```suggestion
- You want to utilize newer language standards, such as C++ 17
```

- You need libraries beyond what the system Python offers
- You want to manage packages from languages other than Python
  in the same space

Workflow differentiators
========================

Some questions to consider as you determine your preferred
workflow and virtual environment: 

- Is your environment shared across multiple code projects?
- Does your environment live alongside your code or in a separate place?
- Do your install steps involve installing any external libraries?

> - Do you want to ship your environment as a speckle of some sort

Reviewer: ```suggestion
- Do you want to ship your environment as an archive of some sort containing the actual files of the environment?
```

  of a big blob of actual stuff?

> - Is a dependency solver a good thing or a bad thing?

Reviewer: This is cheeky.  The point here is that right now, pip is often carelessly clobbering things.  That makes the newest thing installed work, but breaks older things.  If you use a solver, on the other hand, then you may not even be able to install the new thing because the solver knows that the new and old things are incompatible.  Some people rely on pip's breaking behavior, and there a dependency solver is a "bad thing," where really the correct idea is that solvers are good things, and they make sure we keep our environments all happy and hunky-dory.

Author: I removed that bullet point just to keep it clean.

Author: I removed that bullet point.

|

Other conda environments usually contain the same subdirectories
as the default environment.

Virtual environments
====================

A virtual environment is a tool that helps to
keep dependencies required by different projects

> separate by creating isolated Python virtual environments

Reviewer: ```suggestion
separate by creating isolated spaces for them that contain per-project dependencies.
```

for them.

Users can create virtual environments 
using one of several tools such as
Pipenv or Poetry, or a conda virtual
environment. Pipenv and Poetry are based around Python's
built-in venv library, whereas conda has its own notion of virtual
environments that is lower-level (Python itself is a dependency provided
in conda environments).

Some other traits are:

.. list-table::
   :widths: 20 40 40
   :header-rows: 1

   * - 
     - Python virtual environment
     - Conda virtual environment
   * - **Libraries**
     - Statically link, vendor libraries in wheels,
       or use apt/yum/brew/etc.
     - Install system-level libraries as conda dependencies
   * - **System**
     - Depend on base system install of Python
     - Python is independent from system
   * - **Extending environment**
     - Extend environment with pip
     - Extended environment with conda or pip
   * - **Non-Python dependencies**

>      -

Reviewer: extra dash here?

Reviewer: oh wait, sorry, it's because there is no pip equivalent.

Author: So no changes?

Reviewer: Yeah, no changes here.
